### PR TITLE
Must not provide any credentials via HTTP headers

### LINKED ISSUE #188: Clarify that we don't support cookies

### PR DESCRIPTION
**Pull request by vasilvv**
Fixes #188.

***
<a href="https://pr-preview.s3.amazonaws.com/vasilvv/web-transport/pull/312.html" title="Last updated on Nov 10, 2021, 6:56 PM UTC (0f08c4a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webtransport/312/94d996e...vasilvv:0f08c4a.html" title="Last updated on Nov 10, 2021, 6:56 PM UTC (0f08c4a)">Diff</a>

### CODE REVIEW

**Comment by ricea** [comment]
We should probably also disallow client certificates (does QUIC even do those?)

**Comment by yutakahirano** [comment]
> We should probably also disallow client certificates (does QUIC even do those?)

TLS client certificates are part of credentials.

> __Credentials__ are HTTP cookies, TLS client certificates, and authentication entries (for HTTP authentication)
> https://fetch.spec.whatwg.org/#credentials

Modifying "via HTTP headers" may be good.

**Review by aboba — APPROVED** [review]

File: index.bs [hunk 1]
```diff
@@ -190,6 +190,8 @@ There may be multiple [=WebTransport sessions=] on one [=connection=], when pool
 
 To <dfn for=session>establish</dfn> a [=WebTransport session=], follow [[!WEB-TRANSPORT-HTTP3]]
 [section 3.3](https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3/#section-3.3).
+When establishing a session, the client MUST NOT provide any [=credentials=]
+via HTTP headers.
```
Comment (martinthomson) [inline index.bs:194]: Adam mentioned client certificates.  I would remove "via HTTP headers" and then be more explicit and explain that this means "no cookies", "no authorization", and "no client auth".

**Comment by vasilvv** [comment]
"No client auth" is actually somewhat complicated, since it happens at the different stage from sending a request.  The problem is, I'm not sure Fetch has a well-suited knob for turning off the client auth.

**Comment by martinthomson** [comment]
To be clear, I think that this should say "no client auth" period.  That encompasses all of the things, including the difficult certificate-based client authentication @vasilvv is concerned about.  Prohibiting only a subset of the authentication-related things doesn't do that.  I would prefer to say "no client auth that might be populated by the user agent, which includes HTTP headers, cookies, and client certificates".

**Comment by vasilvv** [comment]
The client certificate is a bit tricky here.  In general, we do want to actively pool HTTP/3 connections together with top-level HTTP traffic; however, if we require WebTransport to be non-credentialed, this would almost always not work, since most fetches (including the top-level ones) *are* credentialed.

Similarly with client certificates: if we already have an existing HTTP/3 connection to a server, do we actually want to avoid it just because it has a client cert associated with it

**Comment by ricea** [comment]
> Similarly with client certificates: if we already have an existing HTTP/3 connection to a server, do we actually want to avoid it just because it has a client cert associated with it

Yes. Ambient authentication is dangerous.

**Comment by martinthomson** [comment]
What Adam said. Confused deputy and all that.

Fact is, browsers already need to be able to control whether connections are credentialed. So this isn't so big of a deal.

File: index.bs (hunk 1 above)
Comment (annevk) [reply index.bs:194]: If this ends up referencing https://fetch.spec.whatwg.org/#credentials removing "via HTTP headers" is probably sufficient.

Comment (yutakahirano) [reply index.bs:194]: +1

Comment (jan-ivar) [inline index.bs:194]:
```suggestion
When establishing a session, the client MUST NOT provide any [=credentials=].
```

**Comment by jan-ivar** [comment]
Made text changed as requested and merged.

**Pull request merged**